## Data Wrangling and Cleaning

#### This notebook illustrates a workflow for acquiring a dataset, identifying issues, and correcting those issues in preparation for use in data analysis.  
#### Full Disclosure: this workflow was derived from a lab project I completed as part of a Coursera/IBM Course in Data Analysis with Python, part of the IBM Data Science Specialization that I completed in 2019.

This dataset is an automotive dataset that resides at the University of California-Irvine, Machine Learning Repository.


In [1]:
import pandas as pd
import matplotlib.pylab as plt

In [3]:
url="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv"
#https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

In [4]:
headers=["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [5]:
df_autos=pd.read_csv(url, names=headers)

In [7]:
df_autos.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Just from the file head it is apparent that there are some issues with this file.  There appear to be question marks in place of missing values in the "normalized-losses" column.  So first off, I will convert any ?'s in the datasest to NaN values, using Numpy.

In [9]:
import numpy as np

In [10]:
df_autos.replace("?", np.nan, inplace=True)
df_autos.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Next I can use built-in Python functions for identifyng missing values in the dataset.  Two useful functions are .isnull() and .notnull().

In [11]:
missing_df = df_autos.isnull()
missing_df.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

Now I can use the dataframe I just created, that contains "True" wherever there is a missing vale, to generate a report of columns and their counts of missing values.

I can create a list of the column names, and use that list in a _for loop_ to iterate through the columns and sum the counts of _True_ (missing value) and _False_ for each.

In [13]:
for column in missing_df.columns.values.tolist():
    print("Column: {}".format(column))
    print (missing_df[column].value_counts())
    print("") 

Column: symboling
False    205
Name: symboling, dtype: int64

Column: normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

Column: make
False    205
Name: make, dtype: int64

Column: fuel-type
False    205
Name: fuel-type, dtype: int64

Column: aspiration
False    205
Name: aspiration, dtype: int64

Column: num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

Column: body-style
False    205
Name: body-style, dtype: int64

Column: drive-wheels
False    205
Name: drive-wheels, dtype: int64

Column: engine-location
False    205
Name: engine-location, dtype: int64

Column: wheel-base
False    205
Name: wheel-base, dtype: int64

Column: length
False    205
Name: length, dtype: int64

Column: width
False    205
Name: width, dtype: int64

Column: height
False    205
Name: height, dtype: int64

Column: curb-weight
False    205
Name: curb-weight, dtype: int64

Column: engine-type
False    205
Name: engine-type, dtype: int64

Column: num-of-cylin

I will start with the "normalized-losses" column where we previously dealt with the question marks by converting them to NaNs.  I will replace those missing values with a calculated mean of the entire column.
<ul>


In [15]:
avg_nl = df_autos["normalized-losses"].astype("float").mean(axis=0)
print("Average of normalized-losses: {}".format(avg_nl))

Average of normalized-losses: 122.0


In [16]:
df_autos["normalized-losses"].replace(np.nan, avg_nl, inplace=True)
df_autos.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               122  alfa-romero       gas        std          two   
1          3               122  alfa-romero       gas        std          two   
2          1               122  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Now the NaN values in "normalized-losses" have been replaced by the column average of 122.
Now I will do the same for "bore", "stroke", "horsepower", and "peak-rpm" columns.

In [17]:
avg_bore = df_autos["bore"].astype("float").mean(axis=0)
print("Average of bore: {}".format(avg_bore))

Average of bore: 3.3297512437810957


In [18]:
avg_str = df_autos["stroke"].astype("float").mean(axis=0)
print("Average of stroke: {}".format(avg_str))

Average of stroke: 3.2554228855721337


In [19]:
avg_hp = df_autos["horsepower"].astype("float").mean(axis=0)
print("Average of horsepower: {}".format(avg_hp))

Average of horsepower: 104.25615763546799


In [20]:
avg_pkrpm = df_autos["peak-rpm"].astype("float").mean(axis=0)
print("Average of peak-rpm: {}".format(avg_pkrpm))

Average of peak-rpm: 5125.369458128079


In [21]:
df_autos["bore"].replace(np.nan, avg_bore, inplace=True)
df_autos["stroke"].replace(np.nan, avg_str, inplace=True)
df_autos["horsepower"].replace(np.nan, avg_hp, inplace=True)
df_autos["peak-rpm"].replace(np.nan, avg_pkrpm, inplace=True)

Now let's regenerate our report of missing values and see what's left to deal with:

In [22]:
missing_df2 = df_autos.isnull()
for column in missing_df2.columns.values.tolist():
    print("Column: {}".format(column))
    print (missing_df2[column].value_counts())
    print("")

Column: symboling
False    205
Name: symboling, dtype: int64

Column: normalized-losses
False    205
Name: normalized-losses, dtype: int64

Column: make
False    205
Name: make, dtype: int64

Column: fuel-type
False    205
Name: fuel-type, dtype: int64

Column: aspiration
False    205
Name: aspiration, dtype: int64

Column: num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

Column: body-style
False    205
Name: body-style, dtype: int64

Column: drive-wheels
False    205
Name: drive-wheels, dtype: int64

Column: engine-location
False    205
Name: engine-location, dtype: int64

Column: wheel-base
False    205
Name: wheel-base, dtype: int64

Column: length
False    205
Name: length, dtype: int64

Column: width
False    205
Name: width, dtype: int64

Column: height
False    205
Name: height, dtype: int64

Column: curb-weight
False    205
Name: curb-weight, dtype: int64

Column: engine-type
False    205
Name: engine-type, dtype: int64

Column: num-of-cylinders
False   

Now we just have missing values in "num-of-doors" and "price".  For "num-of-doors" I will simply replace the missing values with whichever value is the most common in the column.  Let's see which value is the most common:

In [23]:
df_autos['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

114 of the rows have four doors, while only 89 have two doors.  So I will replace the four missing values with "four".

In [24]:
df_autos["num-of-doors"].replace(np.nan, "four", inplace=True)

"Price" is actually the variable I am interested in trying to predict once this dataset is cleaned up.  It won't do me any good to replace missing values of price, and there are only 4 missing values, so I will just drop those rows from the analysis.

In [25]:
df_autos.dropna(subset=["price"], axis=0, inplace=True)

Now, since two rows have been dropped, I will reset the index.

In [28]:
df_autos.reset_index(drop=True, inplace=True)
df_autos.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               122  alfa-romero       gas        std          two   
1          3               122  alfa-romero       gas        std          two   
2          1               122  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Looks great.  I have successfully dealt with the missing values in this dataset using several different methods.  In the next notebook, I will continue the wrangling process by doing the following steps:
    - making sure the data are the correct data type
    - standardizing units
    - normalizing data ranges
    - binning variable for grouped analysis
    - creating dummy variables for categorical features

#### Thank you!